### PR TITLE
CNTRLPLANE-3020: Adopt coreos/stream-metadata-go upstream library

### PR DESCRIPTION
**Pull request by sdminonne**
## What this PR does / why we need it:

Replaces 18 custom CoreOS stream metadata Go types in `support/releaseinfo/releaseinfo.go` with canonical equivalents from `github.com/coreos/stream-metadata-go` v0.4.11. This eliminates duplicated type definitions, gains upstream helper methods (e.g. `GetAMI()`, `URN()`), and lays the groundwork for the dual-stream RHEL NodePool feature.

**Scope**: Pure refactor — no behavioral changes.

Key type mappings:
| Old (HyperShift) | New (upstream) |
|---|---|
| `CoreOSStreamMetadata` | `stream.Stream` |
| `CoreOSArchitecture` | `stream.Arch` |
| `CoreRHCOSImage` | `*rhcos.Extensions` |
| `CoreOSAWSImages` | `*stream.AwsImage` |
| `CoreOSGCPImage` | `*stream.GcpImage` |
| `CoreOSPowerVSImage` | `*stream.SingleObject` |
| `CoreOSKubevirtImages` | `*stream.ContainerImage` |
| `HyperVGen1/HyperVGen2` | `Gen1/Gen2` |
| `SHA256` / `URL` (PowerVS) | `Sha256` / `Url` |

## Which issue(s) this PR fixes:

Fixes [CNTRLPLANE-3020](https://redhat.atlassian.net/browse/CNTRLPLANE-3020)

## Special notes for your reviewer:

- All upstream pointer-typed fields have nil checks added at every access point
- The `Formats` map changed from `map[string]map[string]CoreOSFormat` to `map[string]stream.ImageFormat` — access to disk artifacts changed from `artifact["disk"]` to `artifact.Disk`
- Import ordering was auto-fixed by `make lint-fix`

## Checklist:
- [x] Subject and description added to both, commit and PR.
- [x] Relevant issues have been referenced.
- [ ] This change includes docs.
- [x] This change includes unit tests.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **Bug Fixes**
  * Better nil/empty checks and clearer errors when cloud image or marketplace metadata is missing across AWS, Azure, GCP, OpenStack, KubeVirt, and PowerVS to avoid crashes.

* **Refactor**
  * Unified release metadata to a standardized external format for more consistent image resolution and validation.

* **Tests**
  * Updated and expanded tests to exercise new metadata shapes and additional error scenarios.

### CODE REVIEW

**Review by jparrill — COMMENTED** [review]
Dropped some comments. Thanks!

File: hypershift-operator/controllers/nodepool/aws.go [hunk 1]
```diff
@@ -437,12 +437,11 @@ func getWindowsAMI(region string, specifiedArch string, releaseImage *releaseinf
 	}
 
 	// Access the rhel-coreos-extensions aws-winli data
-	winliData := archData.RHCOS.AWSWinLi
-	if winliData.Regions == nil {
+	if archData.RHELCoreOSExtensions == nil || archData.RHELCoreOSExtensions.AwsWinLi == nil || archData.RHELCoreOSExtensions.AwsWinLi.Regions == nil {
 		return "", fmt.Errorf("no aws-winli regions data found in release image metadata")
 	}
 
-	regionData, exists := winliData.Regions[region]
```
Comment (jparrill) [inline hypershift-operator/controllers/nodepool/aws.go:445]: The triple nil-check is correct, but a single error message for three different failure modes makes debugging harder. Consider splitting:
```go
if archData.RHELCoreOSExtensions == nil {
    return "", fmt.Errorf("no rhel-coreos-extensions data found in release image metadata")
}
if archData.RHELCoreOSExtensions.AwsWinLi == nil || archData.RHELCoreOSExtensions.AwsWinLi.Regions == nil {
    return "", fmt.Errorf("no aws-winli regions data found in release image metadata")
}
```
Also — could you add test cases for `RHELCoreOSExtensions == nil` and `AwsWinLi == nil`? The current tests only cover `Regions == nil`.

File: hypershift-operator/controllers/nodepool/kubevirt/kubevirt.go [hunk 2]
```diff
@@ -50,7 +50,10 @@ func defaultImage(nodePoolArch string, releaseImage *releaseinfo.ReleaseImage) (
 		return "", "", fmt.Errorf("couldn't find OS metadata for architecture %q", archName)
 	}
 
-	containerImage := arch.Images.Kubevirt.DigestRef
+	if arch.Images.KubeVirt == nil {
+		return "", "", fmt.Errorf("no kubevirt image metadata present in release")
+	}
+	containerImage := arch.Images.KubeVirt.DigestRef
 	if containerImage == "" {
```
Comment (jparrill) [inline hypershift-operator/controllers/nodepool/kubevirt/kubevirt.go:57]: Good catch adding this nil guard. Could you add a test case in `TestDefaultImage` where `KubeVirt` is nil? Something like:
```go
{
    name: "When KubeVirt image metadata is nil, it should return error",
    arch: hyperv1.ArchitectureAMD64,
    // Images: stream.Images{} — KubeVirt defaults to nil
}
```
Right now the test only covers the empty `DigestRef` path, not the nil pointer path.

File: hypershift-operator/controllers/nodepool/nodepool_controller.go [hunk 3]
```diff
@@ -745,7 +745,10 @@ func defaultNodePoolAMI(region string, specifiedArch string, releaseImage *relea
 		return "", fmt.Errorf("couldn't find OS metadata for architecture %q", specifiedArch)
 	}
 
-	regionData, hasRegionData := arch.Images.AWS.Regions[region]
+	if arch.Images.Aws == nil {
+		return "", fmt.Errorf("release image metadata has no AWS images")
+	}
+	regionData, hasRegionData := arch.Images.Aws.Regions[region]
 	if !hasRegionData {
```
Comment (jparrill) [inline hypershift-operator/controllers/nodepool/nodepool_controller.go:752]: nit: Could you add a test case for `arch.Images.Aws == nil`? The existing tests always populate Aws via `testutils.InitReleaseImageOrDie` or the fake provider, so this error path is never exercised.

File: hypershift-operator/controllers/nodepool/powervs.go [hunk 4]
```diff
@@ -108,12 +110,16 @@ func (c *CAPI) ibmPowerVSMachineTemplate(templateNameGenerator func(spec any) (s
...
 	arch, foundArch := releaseImage.StreamMetadata.Architectures["ppc64le"]
 	if !foundArch {
 		return nil, "", fmt.Errorf("couldn't find OS metadata for architecture %q", "ppc64le")
 	}
 
+	if arch.Images.PowerVS == nil {
+		return nil, "", fmt.Errorf("release image metadata has no PowerVS images")
+	}
+
 	COSRegion := getImageRegion(region)
```
Comment (jparrill) [inline hypershift-operator/controllers/nodepool/powervs.go:123]: Same pattern — this nil guard is correct but there's no test exercising `PowerVS == nil`. The test in `releaseinfo_test.go:TestReleaseInfoPowerVS` only asserts the happy path against fixture data.

File: hypershift-operator/controllers/nodepool/azure.go [hunk 5]
```diff
@@ -133,33 +133,39 @@ func getAzureMarketplaceMetadata(releaseImage *releaseinfo.ReleaseImage, arch st
 	// Extract marketplace metadata from the RHCOS extensions
 	// Structure: .architectures.<arch>.rhel-coreos-extensions.marketplace.azure.no-purchase-plan
 	// Check for nil safety before accessing nested fields
-	if archData.RHCOS.Marketplace.Azure.NoPurchasePlan.HyperVGen1 == nil &&
-		archData.RHCOS.Marketplace.Azure.NoPurchasePlan.HyperVGen2 == nil {
+	if archData.RHELCoreOSExtensions == nil ||
+		archData.RHELCoreOSExtensions.Marketplace == nil ||
+		archData.RHELCoreOSExtensions.Marketplace.Azure == nil ||
+		archData.RHELCoreOSExtensions.Marketplace.Azure.NoPurchasePlan == nil {
+		return nil, nil // No marketplace data available
```
Comment (jparrill) [inline hypershift-operator/controllers/nodepool/azure.go:140]: The cascading nil check is well done. One question: should `RHELCoreOSExtensions == nil` also be tested explicitly? The current mock helper (`createMockReleaseImage`) always sets it to non-nil.

**Comment by jparrill** [comment]
/retest-required

Now revert it's in place, it should work

**Comment by jparrill** [comment]
as we discussed in the ControlPlane Office hours, this is not failing due to this PR.
/override e2e-aws-4-22
